### ISSUE TITLE
Create section for conditional endorsement triple

### ISSUE THREAD

**Issue by nedmsmith** [ietf-rats-wg/draft-ietf-rats-corim#44]
A section, description and cddl for conditional-reference-series-triple-record and reference-endorsed-record. Possibly, these should be divided into separate CDDL files.

**Comment by nedmsmith** [comment]
We need to discuss whether it should be added by extension or added directly to triples-map.

**Comment by nedmsmith** [comment]
Continuing discussion thread here. 
Note Thomas's musings: https://github.com/ietf-rats-wg/draft-ietf-rats-corim/wiki/Thomas'-musings-on-PR%2349 
Given we rewrite the outer layer of the conditional-reference-series-triple-record as an extended environment map. The crstr might look like:
```
crstr = [ 
   ext-environment-record,
   [ + reference-endorsed-record ] ; series of reference + optional endorsement triples
]
```
```
ext-environment-record = [
  env: environment-map
  rvs: [ + measurement-map ]
]
```
Maybe it's helpful to define a series as a stand-alone triple (to reduce apparent complexity?) as:
```
series-records = [ + cond-end-triple-record ]
cond-end-triple-record = [ 
     refv: measurement-map,
     ? endv: measurement-map
]
```
The semantics of the series-triple are that if the refv measurements match evidence, then accept the endv measurements. But rather than processing all the cond-end-triple-records, series logic means halt processing on the first record where refv matches evidence. Note that it isn't necessarily wrong to assert that for the given refv measurements,
there are endv measurements that make sense. The rule could be applied independent of a specific Target Environment. The verifier might need to work from other triples that are used to construct a list of Target Environments to know how to 
apply the rule, but we're already asking the Verifier to do this sort of thing (as I'll explain later).

There may not be a use case for the series-triple that we know of as it stands and we can assume the triple needs to specify a Target Environment (whether is is by way of `environment-map`, or `ext-environment-map`). 

Additionally, the current set of triples in triples-map implies matching context that isn't explicitly captured in CDDL. 
If we consider `reference-triples` and `endorsed-triples`, we expect they are related if they appear in the same tag. 
But the CDDL doesn't express exactly how. The verifier is expected to relate them if the environment-map is common. 
The Verifier might define CDDL as:
```
ref-end-triples => [ + ref-end-triple-record ] ; an anti-series triple
ref-end-triple-record = [ environment-map, [ + cond-end-triple-record] ]
```
Here, `environment-map` is common to both the reference measurements and the endorsement measurements. 
As is pointed out in the notes on appraisal flow https://github.com/ietf-rats-wg/draft-ietf-rats-corim/wiki/Notes-on-the-appraisal-flow 
the endorsement measurements are optional: "Alongside the environment identifiers, the matched Reference Values MAY also be used in selecting applicable Endorsed Values."
These semantics are captured in CDDL in `cond-end-triple-record`. But in the context of the `ref-end-triples`, the array does not imply "series" matching rules (which terminates at the first matched reference values), instead matching is applied to all the records in the array and if there isn't matching Evidence for all of them, appraisal fails. 
We might refer to this matching semantic as "anti-series" for convenience. 

So we can say the current schema describes an anti-series matching semantic for reference and (optional) endorsed values, based on a common `environment-map`.

The series matching semantic is not that different from anti-series. The SGX use case expects the series measurements are related by a common "extended" environment `ext-environment-map` and matching is series instead of anti-series, 
but otherwise it is quite similar to what CoMID is doing currently.

If CoMID were to make `ref-end-triples` a first class triple, rather than implying that the Verifier should do this, 
then the `ref-end-triples` schema (above), for anti-series, would be quite similar to the `cstr-triple` schema (below), for series.
```
...
cstr-triple => [ + cstr ]
...
cstr = [ ext-environment-record, series-records ]
series-records = [+ cond-end-triple-record]

```

The series matching semantic is of course reasonable and ought to be part of the base set of triples structures.

**Comment by shnwc** [comment]
> Additionally, the current set of triples in triples-map implies matching context that isn't explicitly captured in CDDL.
If we consider reference-triples and endorsed-triples, we expect they are related if they appear in the same tag.
But the CDDL doesn't express exactly how. The verifier is expected to relate them if the environment-map is common.
The Verifier might define CDDL as:
...
> These semantics are captured in CDDL in cond-end-triple-record. But in the context of the ref-end-triples, the array does not imply "series" matching rules (which terminates at the first matched reference values), instead matching is applied to all the records in the array and if there isn't matching Evidence for all of them, appraisal fails.
We might refer to this matching semantic as "anti-series" for convenience.

Very good points!

**Comment by nedmsmith** [comment]
We originally defined reference and endorsed triples as:
```
endorsed-triple-record = [
  environment-map
  [ + measurement-map]
]
reference-triple-record = [
  environment-map
  [ + measurement-map]
]
```
but later changed them to:
```
endorsed-triple-record = [
  environment-map
  measurement-map
]
reference-triple-record = [
  environment-map
  measurement-map
]
```
largely because there wasn't a good way to disambiguate multiple measurements having the same code point. 
The idea was the subject (environment-map) could be used to provide disambiguation. 

What is the reasoning for `measurement-map` in `conditional-endorsement-triple-record` having multiples?

**Comment by thomas-fossati** [comment]
> What is the reasoning for `measurement-map` in `conditional-endorsement-triple-record` having multiples?

The background reasoning is in the [musings](https://github.com/ietf-rats-wg/draft-ietf-rats-corim/wiki/Thomas'-musings-on-PR%2349#multiple-measurements-per-triple).

That said, I am happy with the single measurement per triple, i.e.:

```cddl
conditional-endorsement-triple-record = [
  stateful-environment-map
  ; endorsed values
  [ + measurement-map ]
]
```

**Comment by nedmsmith** [comment]
I'm suggesting the following:
```
conditional-endorsement-series-triple-record = [
  stateful-environment-map
  ; order matters: the first matching record wins
  [ + conditional-series-record ]
]

conditional-series-record = [
  ; reference values to be matched against evidence
  refv: measurement-map
  ; endorsed values that apply in case revf matches
  endv: measurement-map 
]
```
is sufficient for series use cases.

I can't tell if there is a use case for:
```
$$triples-map-extension //= (
  &(conditional-endorsement-triples: 9) =>
  [ + conditional-endorsement-triple-record ]
)

conditional-endorsement-triple-record = [
  stateful-environment-map
  ; endorsed values
  [ + measurement-map ]
]
```
But if so, it should be documented that `[ + measurment-map ]` is disambiguated by `mkey`, , or by a `$raw-value-type-choice`. But otherwise, a singleton `measurement-map` can be disambiguated by a more precise `stateful-environment-map` among several `[ + conditional-endorsement-triple-record ]` records.

**Comment by shnwc** [comment]
This definition does not seem correct to me.
```
stateful-environment-map = {
  environment-map
  [ + measurement-map ]
}
```
And it can be simplified to remove nesting. So one form could be:
```
stateful-environment-record = [
  environment-map
  measurement-map
]
```
Which is the same as `reference-triple-record`.

**Comment by thomas-fossati** [comment]
> I'm suggesting the following:
> 
> ```
> conditional-endorsement-series-triple-record = [
>   stateful-environment-map
>   ; order matters: the first matching record wins
>   [ + conditional-series-record ]
> ]
> 
> conditional-series-record = [
>   ; reference values to be matched against evidence
>   refv: measurement-map
>   ; endorsed values that apply in case revf matches
>   endv: measurement-map 
> ]
> ```
> 
> is sufficient for series use cases.

👍 

> I can't tell if there is a use case for:
> 
> ```
> $$triples-map-extension //= (
>   &(conditional-endorsement-triples: 9) =>
>   [ + conditional-endorsement-triple-record ]
> )
> 
> conditional-endorsement-triple-record = [
>   stateful-environment-map
>   ; endorsed values
>   [ + measurement-map ]
> ]
> ```

neither do I :-)  I do have a use case for:

```cddl
conditional-endorsement-triple-record = [
  stateful-environment-map
  ; endorsed values
  measurement-map
]
```

though (i.e., certification data).

Let's drop my `[ + measurement-map ]` and keep the good old `measurement-map`.

**Comment by thomas-fossati** [comment]
> This definition does not seem correct to me.

why?

**Comment by nedmsmith** [comment]
I think the same caveat applies as described above for `conditional-endorsements-triple-record`. Namely, it should be documented that multiple like measurements in `[ + measurment-map ]` in `stateful-environment-map` is disambiguated by `mkey`, or by a `$raw-value-type-choice`. 
But otherwise, a singleton `measurement-map` can be disambiguated by a more precise `environment-map` in `stateful-environment-map` among several `[ + conditional-endorsement-series-triple-record ]` or `[ + conditional-endorsement-triple-record ]` records.

**Comment by nedmsmith** [comment]
I don't know if we have advise for when to use one form of disambiguation over another.

**Comment by thomas-fossati** [comment]
> I think the same caveat applies as described above for `conditional-endorsements-triple-record`. Namely, it should be documented that multiple like measurements in `[ + measurment-map ]` in `stateful-environment-map` is disambiguated by `mkey`, or by a `$raw-value-type-choice`.

WFM

**Comment by thomas-fossati** [comment]
> I don't know if we have advise for when to use one form of disambiguation over another.

the choice may be profile specific

**Comment by nedmsmith** [comment]
> the choice may be profile specific

But there could be advice that a profile author applies to arrive at a conclusion. Or we may not have advice in which case it is a stylistic preference.

**Comment by thomas-fossati** [comment]
I think what we can reasonably say is just "it MUST be possible to disambiguate among measurements of the same type." and add "E.g., via mkey or ..."

**Comment by shnwc** [comment]
> > This definition does not seem correct to me.
> 
> why?

In my test, the tool `diag2cbor` can't parse `.diag` file with data in this `stateful-environment-map`

**Comment by thomas-fossati** [comment]
> > > This definition does not seem correct to me.
> > 
> > why?
> 
> In my test, the tool `diag2cbor` can't parse `.diag` file with data in this `stateful-environment-map`

Not sure what could be wrong.  Can you provide the code?

**Comment by thomas-fossati** [comment]
> Isn't it the requirement that each item in a map must be a name:value pair?

Ah yes, sorry I didn't notice the typo.  Replace `[` with `{` and `]` with `}` to get what I really meant :-)

**Comment by nedmsmith** [comment]
To summarize, the CDDL being proposed is (I think):
```
$$triples-map-extension //= (
  &(conditional-endorsement-series-triples: 8) =>
  [ + conditional-endorsement-series-triple-record ]
)
```
```
conditional-endorsement-series-triple-record = [
  stateful-environment-record,
  ; order matters: the first matching record wins
  [ + conditional-series-record ]
]
```
```
; an environment with a set of measurements that must match evidence
stateful-environment-record = [
  sub: environment-map,
  series: [ + measurement-map ]
]
```
```
conditional-series-record = [
  ; reference values to be matched against evidence
  refv: measurement-map
  ; endorsed values that apply in case revf matches
  endv: measurement-map
]
```
A simple, non-series variant of the "conditional-endorsement-series-triples" could be as follows:

```
$$triples-map-extension //= (
  &(conditional-endorsement-triples: 9) =>
  [ + conditional-endorsement-triple-record ]
)
```
```
conditional-endorsement-triple-record = [
  stateful-environment-record,
  ; endorsed values
  measurement-map
]
```
If there are multiple measurements needed that reuse the same code point. These could be supplied using additional triple record entries. They could be disambiguated using different `stateful-environment-record` values, different `mkey` values, or different `$raw-value-type-choice` values. 

In cases where `authorized-by` is needed, the scope of authorization may need to be clarified. Normally, we expect authorized-by to have triple level scope. That is to say all measurements contained within the same triple record rely on a single authority. If there are multiple sources for measurements, there should be different triple records defined. We don't expect `authorized-by` scope to be finer granularity than the triple record.

Given this is a correct assumption. There only needs to be a single `authorized-by` source for a given triple. That suggests for `conditional-endorsement-series-triples` the `measurement-map` in `stateful-environment-record` could be a reasonable place for `authorized-by`. This also implies the `measurement-map` structures in `conditional-series-record` are unnecessary, at least for the purpose of `authorized-by`. We could replace `measurement-map` with `measurement-values-map` as a way to prevent potentially conflicting authorization being supplied. This also removes `mkey` which suggests any measurements that require disambiguation will need to rely on different `stateful-environment-record` or `$raw-value-type-choice` for that purpose. This is my preference. If there is a use case for keeping `measurement-map` in the series record then there needs to be normative that prevents use of `authorized-by` here.

The `conditional-endorsement-triples` might follow a similar convention depending on the use case requirements. I don't have a strong opinion, but need prose to account for the potential for colliding `authorized-by` statements.

**Comment by shnwc** [comment]
> Consequently, the following CDDL is what I think is the candidate for the PR:
> 
> ```
> $$triples-map-extension //= (
>   &(conditional-endorsement-series-triples: 8) =>
>   [ + conditional-endorsement-series-triple-record ]
> )
> conditional-endorsement-series-triple-record = [
>   stateful-environment-record,
>   ; order matters: the first matching record wins and halts matching
>   [ + conditional-series-record ]
> ]
> ; an environment with a set of measurements that must match evidence
> stateful-environment-record = [
>   environment-map,
>   measurement-map
> ]
> conditional-series-record = [
>   ; reference values to be matched against evidence
>   refv: measurement-values-map
>   ; endorsed values that apply in case revf matches
>   endv: measurement-values-map
> ]
> ```

This CDDL looks good to me.

> We need to discuss whether it should be added by extension or added directly to triples-map.

Adding directly to triples map sounds good, and the CDDL would look like:

```
triples-map = non-empty<{
  ? &(reference-triples: 0) => [ + reference-triple-record ]
  ? &(endorsed-triples: 1)  => [ + endorsed-triple-record ]
  ? &(identity-triples: 2) => [ + identity-triple-record ]
  ? &(attest-key-triples: 3) => [ + attest-key-triple-record ]
  ? &(dependency-triples: 4) => [ + domain-dependency-triple-record ]
  ? &(membership-triples: 5) => [ + domain-membership-triple-record ]
  ? &(coswid-triples: 6) => [ + coswid-triple-record ]
  ? &(conditional-endorsement-series-triples: 8) => [ + conditional-endorsement-series-triple-record ]
  ? &(conditional-endorsement-triples: 9) => [ + conditional-endorsement-triple-record ]
  * $$triples-map-extension
}>

conditional-endorsement-triple-record = [
  stateful-environment-record,
  ; endorsed values
  measurement-values-map
]
```

**Comment by thomas-fossati** [comment]
LGTM!

**Comment by nedmsmith** [comment]
Closed by PR #49
